libnetdata/dictionary/dictionary.c

```diff
@@ -253,78 +582,362 @@ int dictionary_del(DICTIONARY *dict, const char *name) {
     return ret;
 }
 
+// ----------------------------------------------------------------------------
+// traversal with loop
+
+void *dictionary_foreach_start(DICTFE *dfe, DICTIONARY *dict) {
+    if(unlikely(!dfe || !dict)) return NULL;
+
+    dfe->dict = dict;
+    dfe->started_ut = now_realtime_usec();
+
+    dictionary_read_lock(dict);
+
+    NAME_VALUE *nv = dict->first_item;
+    dfe->last_position_index = (void *)nv;
+
+    if(likely(nv)) {
+        dfe->name = nv->name;
+        dfe->value = (void *)nv->value;
+
+        // TODO: increment reference counter on nv;
```

Comment:
Hello @ktsaou ,

Considering that we have the function `DICTIONARY_STATS_SEARCHES_PLUS1` and this `TODO` is here, I arrived in the conclusion that probably you wanted to call it here instead to keep the current `TODO`, please, can you confirm or not my conclusion?

Reply:
The idea of a reference counter is to keep track of the current users of a dictionary item, so that it will not be deleted while it is being used by others.

I added a reference counter maintenance after your comment, but it is far from complete.
So, for the moment, the dictionary cannot keep guarantee that an item cannot be deleted while it is being used by others.